src/Dependencies/Collections/SegmentedList`1.cs

```diff
@@ -124,49 +124,57 @@ public SegmentedList(IEnumerable<T> collection)
         public int Capacity
         {
             get => _items.Length;
-            set
+            set => SetCapacityInternal(value, isExactCapacity: true);
+        }
+
+        private void SetCapacityInternal(int value, bool isExactCapacity)
+        {
+            if (value < _size)
             {
-                if (value < _size)
-                {
-                    ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.value, ExceptionResource.ArgumentOutOfRange_SmallCapacity);
-                }
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.value, ExceptionResource.ArgumentOutOfRange_SmallCapacity);
+            }
 
-                if (value == _items.Length)
-                    return;
+            if (value == _items.Length)
+                return;
 
-                if (value <= 0)
-                {
-                    _items = s_emptyArray;
-                    return;
-                }
+            if (value <= 0)
+            {
+                _items = s_emptyArray;
+                return;
+            }
 
-                if (_items.Length == 0)
-                {
-                    // No data from existing array to reuse, just create a new one.
-                    _items = new SegmentedArray<T>(value);
-                }
-                else
-                {
-                    // Rather than creating a copy of _items, instead reuse as much of it's data as possible.
-                    var segments = SegmentedCollectionsMarshal.AsSegments(_items);
+            if (_items.Length == 0)
+            {
+                // No data from existing array to reuse, just create a new one.
+                _items = new SegmentedArray<T>(value);
+            }
+            else
+            {
+                // Rather than creating a copy of _items, instead reuse as much of it's data as possible.
+                var segments = SegmentedCollectionsMarshal.AsSegments(_items);
 
-                    var oldSegmentCount = segments.Length;
-                    var newSegmentCount = (value + SegmentedArrayHelper.GetSegmentSize<T>() - 1) >> SegmentedArrayHelper.GetSegmentShift<T>();
+                var oldSegmentCount = segments.Length;
+                var newSegmentCount = (value + SegmentedArrayHelper.GetSegmentSize<T>() - 1) >> SegmentedArrayHelper.GetSegmentShift<T>();
 
-                    // Grow the array of segments, if necessary
-                    Array.Resize(ref segments, newSegmentCount);
+                if (newSegmentCount > segments.Length)
+                {
+                    // Grow the array of segments, if necessary. If isExactCapacity is false, then this array may end up
+                    // having null entries at it's end when this method completes. Minimally doubling the outer array size
```

Comment:
>having null entries at it's end when this method completes [](http://example.com/codeflow?start=23&length=58)

@sharwell 

As you mentioned offline, this could be problematic as various code assumes that entries in the array aren't null. Still looking into this, but to me it seems like it may require changing SegmentedArray to indicate that it's inner arrays can be null.

Reply:
Made changes to SegmentedArray _items array to explicitly acknowledge that it may now contain nulls. Commented the _items declaration to indicate which parts of it may be null. By making this nullable, it catches all the locations that I missed that could have null ref'ed with this change.

Reply:
Updated the benchmark numbers with the changes from commit 4. Not a big difference from the numbers taken without the nullability changes, still significantly better in the measured scenarios than without this PR.